core/src/main/java/io/questdb/cairo/map/UnorderedVarcharMap.java

```diff
@@ -0,0 +1,740 @@
+        return sizeAndFlags & 0x3FFFFFFF;
+    }
+
+    long entrySize() {
+        return entrySize;
+    }
+
+    boolean isZeroKey(long startAddress) {
+        long packedHashAndSize = Unsafe.getUnsafe().getLong(startAddress);
+        return packedHashAndSize == 0;
+    }
+
+    class Key implements MapKey {
+        private byte flags;
+        private long ptr;
+        private long ptrWithUnstableFlag;
+        private int size;
+
+        @Override
+        public long commit() {
+            return KEY_SIZE; // we don't need to track the actual key size
+        }
+
+        @Override
+        public void copyFrom(MapKey srcKey) {
+            Key srcVarcharKey = (Key) srcKey;
+            size = srcVarcharKey.size;
+            flags = srcVarcharKey.flags;
+            if ((srcVarcharKey.ptrWithUnstableFlag & PTR_UNSTABLE_MASK) == 0) {
+                // stable pointer
+                assert srcVarcharKey.ptrWithUnstableFlag == srcVarcharKey.ptr;
+                ptrWithUnstableFlag = srcVarcharKey.ptr;
+                ptr = srcVarcharKey.ptr;
+            } else {
+                ptr = allocator.malloc(size);
+                Vect.memcpy(ptr, srcVarcharKey.ptr, size);
+                ptrWithUnstableFlag = ptr | PTR_UNSTABLE_MASK;
+            }
+        }
+
+        @Override
+        public MapValue createValue() {
+            long hash = Hash.hashMem64(ptr, size);
+            return createValue(hash);
+        }
+
+        @Override
+        public MapValue createValue(long hashCode) {
+            long index = hashCode & mask;
+            long startAddress = getStartAddress(index);
+
+            long loadedHashSizeFlags = Unsafe.getUnsafe().getLong(startAddress);
+            long currentHashSizeFlags = packHashSizeFlags(hashCode, size, flags);
+            if (loadedHashSizeFlags == 0) {
+                return asNew(startAddress, hashCode, ptr, ptrWithUnstableFlag, size, currentHashSizeFlags, value);
+            }
+            if (loadedHashSizeFlags == currentHashSizeFlags) {
+                long currentPtr = Unsafe.getUnsafe().getLong(startAddress + 8) & PTR_MASK;
+                if (Vect.memeq(currentPtr, ptr, size)) {
+                    return valueOf(startAddress, false, value);
+                }
+            }
+            return probe0(startAddress, hashCode, ptr, ptrWithUnstableFlag, size, currentHashSizeFlags, value);
+        }
+
+        @Override
+        public MapValue findValue() {
+            return findValue(ptr, size, flags, value);
+        }
+
+        @Override
+        public MapValue findValue2() {
+            return findValue(ptr, size, flags, value2);
+        }
+
+        @Override
+        public MapValue findValue3() {
+            return findValue(ptr, size, flags, value3);
+        }
+
+        @Override
+        public long hash() {
+            return Hash.hashMem64(ptr, size);
+        }
+
+        public Key init() {
+            size = -1;
+            ptrWithUnstableFlag = -1;
```

Comment:
nit: don't we need to reset `ptr` and `flags` here?

Reply:
I think we don't. a left-over, thanks!

Reply:
it now behaves the same as when null is inserted. the behaviour is undefined as discussed. 
perhaps we can get a tiny bit of speed by not doing anything at all? wdyt?

Reply:
Yes, it's fine if we don't reset the fields here. Worth putting a clarifying comment though.

Reply:
alright, let me remove the initialization code

Reply:
@puzpuzpuz: done